src/cached_pool/mod.rs

```diff
@@ -0,0 +1,640 @@
+            self.registry.deregister_and_await_termination(self.id);
+        } else {
+            self.registry.deregister(self.id);
+        }
+    }
+}
+
+/// Configuration for `CachedPool`.
+#[derive(Clone, Copy, Debug)]
+pub struct Config {
+    /// The maximum number of threads the thread pool can spawn at the same time.
+    pub max_threads: u32,
+    /// The minimum number of threads the thread pool will contain when idle.
+    pub min_threads: u32,
+    /// The amount of time a thread can be idle before expiring.
+    pub max_idle: Duration,
+}
+
+impl Config {
+    /// Creates a new `Config` with the specified max idle time for threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn max_idle(max_idle: Duration) -> Config {
+        Config { max_idle: max_idle, ..Config::default() }
+    }
+
+    /// Creates a new `Config` with the specified maximum number of threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn max_threads(max_threads: u32) -> Config {
+        Config { max_threads: max_threads, ..Config::default() }
+    }
+
+    /// Creates a new `Config` with the specified minimum number of threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn min_threads(min_threads: u32) -> Config {
+        Config { min_threads: min_threads, ..Config::default() }
+    }
+
+    #[inline]
+    fn max_idle_ms(&self) -> u64 {
+        self.max_idle.as_millis().expect("Overflowed converting Duration to milliseconds!")
+    }
+}
+
+impl Default for Config {
+    #[inline]
+    fn default() -> Self {
+        use std::u32;
+        Config {
+            max_threads: u32::MAX,
+            min_threads: 0,
+            max_idle: Duration::from_secs(5 * 60),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct Registry {
+    tx: mio::Sender<EventLoopAction>,
+    count: Option<Arc<()>>,
+}
+
+impl Registry {
+    /// Create a new thread pool with the given maximum number of threads
+    /// and maximum idle time before thread expiration.
+    fn new() -> Registry {
+        let mut config = EventLoopConfig::default();
+        config.notify_capacity(1_000);
+        let mut event_loop = EventLoop::configured(config).expect(pos!());
+        let tx = event_loop.channel();
+        thread::Builder::new()
+            .name("CachedPool Registry".to_string())
+            .spawn(move || {
+                if let Err(e) = event_loop.run(&mut Dispatcher::new()) {
+                    error!("Dispatcher: event loop failed, {:?}", e);
+                }
+            })
+            .expect(pos!());
+        Registry {
+            tx: tx,
+            count: Some(Arc::new(())),
+        }
+    }
+
+    /// Submit a new task to a thread.
+    ///
+    /// Fails if all threads are busy with tasks, and the thread pool
+    /// is running its maximum configured number of threads.
+    fn execute<F>(&self, pool_id: PoolId, f: F) -> Result<(), Box<Task + Send>>
```

Comment:
`mio::Sender::send` can fail if mio's backing queue (which is fixed-size) is full. I don't want to expose this in the API, and I intend to replace mio's backing queue with a custom queue that grows indefinitely, once such an option is available. At that point, it will be safe to call unwrap. Until that point, these can technically panic, but it's unlikely.